src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/helper/builders/CourseActivity.kt

```diff
@@ -0,0 +1,220 @@
+    CourseActivity(
+      courseActivityId = courseActivityId,
+      code = code,
+      program = programService,
+      caseloadId = prisonId,
+      prison = lookupAgency(prisonId),
+      description = description,
+      capacity = capacity,
+      active = active,
+      scheduleStartDate = LocalDate.parse(startDate),
+      scheduleEndDate = endDate?.let { LocalDate.parse(it) },
+      iepLevel = lookupIepLevel(minimumIncentiveLevelCode),
+      internalLocation = internalLocationId?.let { lookupAgencyInternalLocation(it, prisonId) },
+      excludeBankHolidays = excludeBankHolidays,
+    )
+      .let { save(it) }
+      .also { courseActivity = it }
+
+  override fun payRate(
+    iepLevelCode: String,
+    payBandCode: String,
+    startDate: String,
+    endDate: String?,
+    halfDayRate: Double,
+  ) =
+    courseActivityPayRateBuilderFactory.builder().build(
+      courseActivity,
+      iepLevelCode,
+      payBandCode,
+      startDate,
+      endDate,
+      halfDayRate,
+    )
+      .also { courseActivity.payRates += it }
+
+  override fun courseSchedule(
+    courseScheduleId: Long,
+    scheduleDate: String,
+    startTime: String,
+    endTime: String,
+    slotCategory: SlotCategory,
+    scheduleStatus: String,
+  ) = courseScheduleBuilderFactory.builder().build(
+    courseActivity,
+    courseScheduleId,
+    scheduleDate,
+    startTime,
+    endTime,
+    slotCategory,
+    scheduleStatus,
+  )
+    .also {
+      courseActivity.courseSchedules += it
+    }
+
+  override fun courseScheduleRule(
+    id: Long,
+    startTimeHours: Int,
+    startTimeMinutes: Int,
+    endTimeHours: Int,
+    endTimeMinutes: Int,
+    monday: Boolean,
+    tuesday: Boolean,
+    wednesday: Boolean,
+    thursday: Boolean,
+    friday: Boolean,
+    saturday: Boolean,
+    sunday: Boolean,
+  ) = courseScheduleRuleBuilderFactory.builder().build(
+    courseActivity,
+    id,
+    startTimeHours,
+    startTimeMinutes,
+    endTimeHours,
+    endTimeMinutes,
+    monday,
+    tuesday,
+    wednesday,
+    thursday,
+    friday,
+    saturday,
+    sunday,
+  )
+    .also {
+      courseActivity.courseScheduleRules += it
+    }
+
+  fun save(courseActivity: CourseActivity) = repository?.save(courseActivity)
+    ?: courseActivity
+  fun lookupAgency(id: String): AgencyLocation = repository?.lookupAgency(id)
+    ?: AgencyLocation(id = id, description = id, type = AgencyLocationType.PRISON_TYPE, active = true)
+  fun lookupIepLevel(code: String): IEPLevel = repository?.lookupIepLevel(code)
+    ?: IEPLevel(code = code, description = code)
+  fun lookupAgencyInternalLocation(internalLocationId: Long, prisonId: String) = repository?.lookupAgencyInternalLocation(internalLocationId)
+    ?: AgencyInternalLocation(
+      locationId = internalLocationId,
+      active = true,
+      locationType = "CLAS",
+      agencyId = prisonId,
+      description = "Classroom 1",
+      locationCode = internalLocationId.toString(),
+    )
```

Comment:
The repository is nullable so that we can use the DSL for unit tests where we have to manually create an instance and no repositories are available.
In unit tests we just construct the thing that would normally either lived or be saved to the database.